lte/gateway/c/core/oai/include/map.h

```diff
@@ -0,0 +1,232 @@
+/**
+ * Copyright 2020 The Magma Authors.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+#include <pthread.h>
+
+#include <iostream>
+#include <unordered_map>
+#include <mutex>
+#include <string>
+
+#include "lte/gateway/c/core/oai/lib/3gpp/3gpp_23.003.h"
+
+namespace std {
+typedef size_t hash_size_t;
+/*Default hash function*/
+static hash_size_t def_hashfunc(const void* const keyP, const int key_sizeP) {
+  hash_size_t hash = 0;
+  int key_size     = key_sizeP;
+
+  // may use MD4 ?
+  while (key_size > 0) {
+    uint32_t val = 0;
+    int size     = sizeof(val);
+    while ((size > 0) && (key_size > 0)) {
+      val = val << 8;
+      val |= (reinterpret_cast<const uint8_t*>(keyP))[key_size - 1];
+      size--;
+      key_size--;
+    }
+    hash ^= val;
+  }
+
+  return hash;
+}
+// specialise std::equal_to function for type guti_m5_t
+template<>
+struct equal_to<guti_m5_t> {
+  bool operator()(const guti_m5_t& guti1, const guti_m5_t& guti2) const {
+    return (guti1.m_tmsi == guti2.m_tmsi);
+  }
+};
+
+// specialise std::hash function for type guti_m5_t
+// HashFunction for Guti as key and unit64 data
+// Note: This HashFunc in turn calls the def_hashfunc which supports hashing
+// only for unit64
+template<>
+struct hash<guti_m5_t> {
+  size_t operator()(const guti_m5_t& k) const {
+    return def_hashfunc(&k, sizeof(k));
+  }
+};
+}  // namespace std
+namespace magma5g {
```

Comment:
can we use `magma` namespace here? since this will be reused for hashtables on non-5g code as well

Reply:
This change has been addressed.